rbf-chunker/src/main.rs

```diff
@@ -0,0 +1,211 @@
+    packet_offset: usize,
+    packet_size: usize,
+    discards: usize,
+    target_crc: [u8; 2],
+    crc: [u8; 2],
+}
+
+impl PacketInfo {
+    fn new(
+        packet_number: u16,
+        packet_offset: usize,
+        discards: usize,
+        packet: &[u8],
+        crc: u16,
+    ) -> Self {
+        PacketInfo {
+            packet_number,
+            packet_offset,
+            discards,
+            packet_size: packet.len(),
+            target_crc: array_ref!(packet, packet.len() - 2, 2).clone(),
+            crc: [(crc & 0xFF) as u8, ((crc & !0xFF) >> 8) as u8],
+        }
+    }
+}
+
+fn dump_packet_infos(target_dir: &PathBuf, packet_infos: &[PacketInfo]) -> std::io::Result<()> {
+    let mut crc_info_file = File::create(target_dir.join("packet_infos"))?;
+    for crc_info in packet_infos {
+        writeln!(&mut crc_info_file, "{:?}", crc_info)?;
+    }
+    Ok(())
+}
+
+fn handle_file(target_dir: PathBuf, filename: PathBuf) -> std::io::Result<()> {
+    const PACKET_MIN_SIZE: usize = 300;
+    const PACKET_MAX_SIZE: usize = 1024 * 2;
+    const DISCARD_MAX_SIZE: usize = 1024 * 5;
+    const HEADER_SIZE: usize = 0x84;
+
+    if target_dir.exists() {
+        Err(Error::new(
+            ErrorKind::AlreadyExists,
+            "Target dir already exists. Refuse to overwrite.",
+        ))?
+    }
+
+    std::fs::create_dir(target_dir.clone())?;
+    let mut file = File::open(filename.clone())?;
+    let mut working_buffer: ArrayGen<u8, { PACKET_MAX_SIZE + DISCARD_MAX_SIZE }> =
+        [0; PACKET_MAX_SIZE + DISCARD_MAX_SIZE].into();
+
+    // Read the header.
+    file.read_exact(&mut working_buffer[..HEADER_SIZE])?;
+    File::create(target_dir.join("header"))?.write_all(&working_buffer[..HEADER_SIZE])?;
+
+    let mut running_crc = crc16::State::<crc16::MODBUS>::new();
+    let mut packet_infos = vec![];
+
+    let mut read_so_far = 0;
+    let mut discard_so_far = 0;
+    let mut packet_offset = HEADER_SIZE;
+    let mut packet_number = 0u16;
+    loop {
+        match file.read(
+            &mut working_buffer[discard_so_far + read_so_far..discard_so_far + PACKET_MAX_SIZE],
+        ) {
+            Ok(0) => {
+                if read_so_far == 0 {
+                    // Last packet was actually the footer.
+                    // Discard it.
+                    //
+                    // Wtf???
+                    eprintln!("{:?}: Discarding last packet as was footer.", filename);
+
+                    assert!(packet_number > 0);
+                    let filename = format!("packet{:05}", packet_number - 1);
+                    std::fs::remove_file(target_dir.join(filename))?;
+                    packet_infos.pop().unwrap();
+                }
+                File::create(target_dir.join("footer"))?
+                    .write_all(&working_buffer[..discard_so_far + read_so_far])?;
+                dump_packet_infos(&target_dir, &packet_infos)?;
+
+                return Ok(());
+            }
+            Ok(n) => {
```

Comment:
This entire thing should be its own function; we'll need it for far more than just chunking files.